### PR TITLE
Make snapshots usable again

### LINKED ISSUE #1218: Snapshot feature will modify existing snapshots without passing `--update-snapshots` flag
## Description

The results with snapshots are inconsistent when ran multiple times with no code changes. I've provided a repository with an example below.

Upon cloning the repo and running `npm test` it will pass the tests but leave the snapshots modified (See output of 2nd pass below). If you run tests again, it will throw an error (See 3rd pass output below).

For what it's worth, I was using the `master` branch after the snapshot feature first landed and had no issues. It seems like the bug was introduced when the magic assert feature was introduced. I even tried deleting the snapshot folder and getting a fresh run, but it still produces the same results.

### Error Message & Stack Trace

```sh
# 2nd pass
> ava

  4 passed

~/Projects/ava-snapshot-issue master*

# 3rd pass
> ava

  3 passed
  1 failed

  renders custom width
  test.js:9

   8:
   9:   t.snapshot(tree)
   10: }

  Error: Unexpected token m in JSON at position 28

  Test.snapshot (test.js:9:5)

npm ERR! Test failed.  See above for more details.
```

### Config

Copy the relevant section from `package.json`:

```json
{
  "ava": {
    "babel": "inherit",
    "require": "babel-register"
  }
}
```

### Command-Line Arguments

Copy your npm build scripts or the `ava` command used:

```
npm test
```

## Relevant Links

See reproducable repo here: https://github.com/wyze/ava-snapshot-issue

## Environment

Node.js v7.1.0
darwin 16.1.0
npm 4.0.3
AVA 0.18

### LINKED ISSUE #1220: Snapshot feature shows empty diff when changing order of keys on object
## Description

When using the snapshot feature, if the new value is an object that is identical to the snapshot **but with a different key order**, the diff shown is **empty**.

### Test Source

Before:

```js
	...
	t.snapshot({
		coolNumber: 5 + 8,
		otherNumber: 9 + 8,
	});
	...
```

After altering the test so that the keys are in reversed order:
```js
	...
	t.snapshot({
		otherNumber: 9 + 8,
		coolNumber: 5 + 8,
	});
	...
```

### Error Message & Stack Trace

```
  1 failed

  banana
  test/test.js:6

   5: test('banana', async t => {
   6:   t.snapshot({             
   7:     otherNumber: 9 + 8,    

  Difference:

 
  Please check your code or --update-snapshots

  Test.<anonymous> (test/test.js:6:4)
  Test.fn (test/test.js:5:1)
```

### Config

package.json:
```json
{
  "ava": {
    "require": [
      "babel-register"
    ],
    "babel": "inherit"
  }
}
```

.babelrc:
```json
{
    "plugins": [
        "transform-runtime"
    ],
    "presets": [
        "stage-0",
        "es2015-node4"
    ]
}
```

### Command-Line Arguments

```
ava
```

## Environment

```
Node.js v4.6.2
linux 4.8.0-34-generic

AVA 0.18.0

npm 3.10.9
```

### LINKED ISSUE #1254: Snapshot is doubly-escaped / hard to see diffs
The example from the readme gives:

```js
exports[`HelloWorld component 1`] = `"\"\\n<h1>\\n    Hello World...!\\n</h1>\\n\""`;
```

(looks like it's json-stringified, with extra `"` quotes)

I'd love to use the snapshot feature, but non-diffable snapshots are a no-go. Could the exported value not just be a json object instead of a stringified json object?

### PR DESCRIPTION
**Pull request by novemberborn**
Had to start with another refactor, this time of `Runner`.

Anyway, the cool part is making snapshots work again:

* Make the Runner manage the snapshot state. Thread an accessor to the
`t.snapshot()` assertion.

* Save snapshots when runner has finished. Fixes #1218.

* Use jest-snapshot directly, without serializing values. Use jest-diff
to generate the diff if the snapshot doesn't match. This does mean the
output is not colored and has other subtle differences with how other
assertions format values, but that's OK for now. Fixes #1220, fixes #1254.

* Pin jest-snapshot and jest-diff versions. This isn't ideal but we're
using private APIs and making other assumptions, so I'm not comfortable
with using a loose SemVer range.

It's still not ideal (see #1275), especially since the diff output is different from other assertions, but at least it makes snapshots usable.

### CODE REVIEW

File: test/fixture/assert.snap [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (sindresorhus) [inline test/fixture/assert.snap]: We need to change this to `AVA Snapshot v1`. We don't want to mislead people with being fully compatible with Jest and it's also confusing for AVA users that it doesn't say AVA.

We might need to open an issue on Jest about getting an option for this, but for now I guess we can just read the file in, modify it, and write it out again.

**Comment by sindresorhus** [comment]
This looks very good. I almost couldn't tell the difference with our diff. Usable now is better than perfect at some point.

File: test/fixture/assert.snap (hunk 1 above)
Comment (novemberborn) [inline test/fixture/assert.snap]: I think `jest-snapshot` just treats these files as CommonJS modules so rewriting the header should be harmless. Will do.

File: lib/runner.js [hunk 2]
```diff
@@ -95,100 +135,55 @@ class Runner extends EventEmitter {
...
 			return Promise.resolve(null);
 		}
 
-		this.tests.on('test', this._addTestResult);
-
 		this.hasStarted = true;
-
-		return Promise.resolve(this.tests.build().run()).then(this._buildStats);
+		this.tests.on('test', result => this.addTestResult(result));
+		return new Promise(resolve => {
```
Comment (novemberborn) [inline lib/runner.js]: Ah `Promise` here is indeed from `bluebird`. I guess I'll rename it to `Bluebird.try()`.

File: lib/runner.js [hunk 3]
```diff
@@ -95,100 +135,55 @@ class Runner extends EventEmitter {
...
 		if (options.runOnlyExclusive && !this.tests.hasExclusive) {
 			return Promise.resolve(null);
 		}
 
-		this.tests.on('test', this._addTestResult);
-
 		this.hasStarted = true;
-
-		return Promise.resolve(this.tests.build().run()).then(this._buildStats);
+		this.tests.on('test', result => this.addTestResult(result));
```
Comment (novemberborn) [inline lib/runner.js]: Sure.

File: test/fixture/assert.snap (hunk 1 above)
Comment (novemberborn) [inline test/fixture/assert.snap]: Actually no, `jest-snapshot` requires snapshot files to start with this header: https://github.com/facebook/jest/blob/38ca942f6f2979ed25f54510173dbd2761dd7e04/packages/jest-snapshot/src/utils.js#L37

We could add a comment after but that becomes annoyingly complex.

Comment (sindresorhus) [inline test/fixture/assert.snap]: @novemberborn Can you open an issue on Jest to make it configurable? They seem open to making it more generic when needed.

I'm happy to land this as is for now, but long-term we'll want our own naming and versioning.

Comment (novemberborn) [inline test/fixture/assert.snap]: > Can you open an issue on Jest to make it configurable? They seem open to making it more generic when needed.

I don't know, if you look at the [error messages it very much assumes running inside Jest](https://github.com/facebook/jest/blob/38ca942f6f2979ed25f54510173dbd2761dd7e04/packages/jest-snapshot/src/utils.js#L40:L76).

> I'm happy to land this as is for now, but long-term we'll want our own naming and versioning.

I think we want our own alternative, really.

Comment (sindresorhus) [inline test/fixture/assert.snap]: Too bad, would have been nice to reuse that code.

// @cpojer

**Comment by cpojer** [comment]
They are still Jest snapshots though, aren't they? If we can, we should keep the header because it looks to me like they are compatible. We can update error messages and the documentation link that it links to, though. Feel free to send PRs to Jest to make the public API more usable as we'll undoubtedly break private APIs in the future.

**Pull request merged**